frontend/catalyst/pennylane_extensions.py

```diff
@@ -191,12 +191,22 @@ def _make_jaxpr_check_differentiable(f: Differentiable, grad_params: GradParams,
                     return_ops.append(eq.primitive)
                     break
 
-        if method == "ps" and any(prim not in [expval_p, probs_p] for prim in return_ops):
+        assert isinstance(
+            f, qml.QNode
+        ), "Differentiation methods other than finite-differences can only operate on a QNode"
+        if f.diff_method is None:
+            raise ValueError(
```

Comment:
Minor: we use `TypeError`s below but a `ValueError` here, not sure which is more appropriate though.

Reply:
The Python docs say `TypeError` means "inappropriate argument type" while `ValueError` means "inappropriate argument value (of correct type)". I originally returned a ValueError because my thinking was "a QNode marked non-differentiable is still a QNode".

However, from a compiler's perspective, this function seems to be doing type checking such that returning a value of a wrong type constitutes a TypeError. Given that the other two cases include the `diff_method` as part of a QNode's *type*, it feels appropriate that trying to differentiate a QNode marked non-differentiable would be a type error.

Reply:
(flyby comment): I would say best practice typically, for a broader new category of error, is that libraries should subclass the closest Python error and create their own (e.g.,

```python
class GradientTypeError(TypeError):
    pass
```

The reason this is considered best practice is it allows other developers to try-except on _just_ the more specific `GradientTypeError`.